Update snowglobes.py

jpkneller opened the pull request:
1) Added get_transformation function to return a flavor transformation class from a string 2) Added get_all_models_dict function to build a master dictionary of models 3) Modified the two generate_* functions to allow a user to pass instances of a flavor transformation as well as a string, and use the get_all_models_dict to identify models

jpkneller committed:
Update snowglobes.py

1) Added get_transformation function to return a flavor transformation class from a string
2) Added get_all_models_dict function to build a master dictionary of models
3) Modified the two generate_* functions to allow a user to pass instances of a flavor transformation as well as a string, and use the get_all_models_dict to identify models

jpkneller committed:
Update snowglobes.py

JostMigenda reviewed (commented):
Good idea; just a couple of comments on the implementation details.

jpkneller commented on python/snewpy/snowglobes.py:
```diff
@@ -38,7 +40,63 @@
+                                  'TwoFlavorDecoherence_NMO': TwoFlavorDecoherence(NMO_mix_params), 
+                                  'TwoFlavorDecoherence_IMO': TwoFlavorDecoherence(IMO_mix_params), 
+                                  'ThreeFlavorDecoherence': ThreeFlavorDecoherence()
+                                  }
```
Yes. I know some older codes which use SNEWPY will now fail but a) I doubt many (any) people used these prescriptions and b) NeutrinoDecay_NMO, NeutrinoDecay_IMO, QuantumDecoherence_NMO and QuantumDecoherence_IMO were examples of what we now call a Transformation Chain and I want to force users into using that explicitly. The TwoFlavorDecoherence in the past version was only for NMO; CompleteExchange was missing.

jpkneller replied:
The SNEWPY2 paper draft does this say this repeatedly. I agree we should add a deprecation warning if the user passes strings to either generate_* function: we want to get away from dictionaries because they are so limiting.

JostMigenda committed:
implement most comments from code review

JostMigenda committed:
fix str.find bug; add error handling; clean-up

JostMigenda committed:
Merge branch 'main' into jpkneller---new-generate_

JostMigenda committed:
add DeprecationWarning when giving flavor_transformation as a string

fix warn

JostMigenda committed:
add required argument to ThreeFlavorDecoherence

JostMigenda committed:
restore NeutrinoDecay and QuantumDecoherence support
Includes improved warning about using default parameters for both transformations.

fix ND/QD

jpkneller commented:
> If anything, I wonder if we should remove the three entries (`CompleteExchange`, `TwoFlavorDecoherence_NMO` and `TwoFlavorDecoherence_IMO`) that were newly added in this PR. They are not needed for backward compatibility, so it would be okay to require users to specify those as `FlavorTransformation` objects only …

I am okay either way. As you said, the new entries are not for backward compatibility, they are just for completeness. I would keep them only because it would seem strange to have some flavor transformations which are not accessible that way.

jpkneller reviewed (commented):
Looks good to me but I can't approve.

sybenzvi committed:
Merge branch 'main' into jpkneller---new-generate_

sybenzvi reviewed (approved):
Discussion on 2026-06-09 (SNEWS 2.0 signal prediction call): While we don't love the fragility of passing strings, to maintain backwards compatibility, we are approving the PR.
